pkg/controller/direct/firestore/firestorefield_controller.go

```diff
@@ -0,0 +1,396 @@
+	}
+
+	if err := common.NormalizeReferences(ctx, reader, obj, nil); err != nil {
+		return nil, fmt.Errorf("normalizing references: %w", err)
+	}
+
+	id, err := obj.GetIdentity(ctx, reader)
+	if err != nil {
+		return nil, err
+	}
+
+	var desired *pb.Field
+	{
+		mapCtx := &direct.MapContext{}
+		desired = FirestoreFieldSpec_v1alpha1_ToProto(mapCtx, &obj.Spec)
+		if mapCtx.Err() != nil {
+			return nil, mapCtx.Err()
+		}
+	}
+
+	return &firestoreFieldAdapter{
+		id:                   id.(*krm.FirestoreFieldIdentity),
+		firestoreAdminClient: firestoreAdminClient,
+		desired:              desired,
+	}, nil
+}
+
+func (m *firestoreFieldModel) AdapterForURL(ctx context.Context, url string) (directbase.Adapter, error) {
+	// The url format should match the Cloud-Asset-Inventory format: https://cloud.google.com/asset-inventory/docs/resource-name-format
+	if !strings.HasPrefix(url, "//firestore.googleapis.com/") {
+		return nil, nil
+	}
+
+	url = strings.TrimPrefix(url, "//firestore.googleapis.com/")
+
+	id := &krm.FirestoreFieldIdentity{}
+	if err := id.FromExternal(url); err != nil {
+		// Not recognized
+		return nil, nil
+	}
+
+	firestoreAdminClient, err := newFirestoreAdminClient(ctx, m.config)
+	if err != nil {
+		return nil, err
+	}
+	return &firestoreFieldAdapter{
+		id:                   id,
+		firestoreAdminClient: firestoreAdminClient,
+	}, nil
+}
+
+type firestoreFieldAdapter struct {
+	id                   *krm.FirestoreFieldIdentity
+	firestoreAdminClient *apiv1.FirestoreAdminClient
+	desired              *pb.Field
+	actual               *pb.Field
+}
+
+var _ directbase.Adapter = &firestoreFieldAdapter{}
+
+func (a *firestoreFieldAdapter) Find(ctx context.Context) (bool, error) {
+	log := klog.FromContext(ctx)
+
+	if a.id == nil {
+		return false, nil
+	}
+
+	fqn := a.id.String()
+	log.V(0).Info("getting FirestoreField", "name", fqn)
+
+	req := &pb.GetFieldRequest{Name: fqn}
+	firestorefieldpb, err := a.firestoreAdminClient.GetField(ctx, req)
+	if err != nil {
+		if direct.IsNotFound(err) {
+			return false, nil
+		}
+		return false, fmt.Errorf("getting FirestoreField %q: %w", fqn, err)
+	}
+
+	a.actual = firestorefieldpb
+	return true, nil
+}
+
+func (a *firestoreFieldAdapter) Create(ctx context.Context, createOp *directbase.CreateOperation) error {
+	fqn := a.id.String()
+
+	log := klog.FromContext(ctx)
+	log.V(0).Info("creating FirestoreField", "name", fqn)
+
+	req := &pb.UpdateFieldRequest{
+		Field: direct.ProtoClone(a.desired),
+	}
+	req.Field.Name = fqn
+
+	req.UpdateMask = a.allFields()
+
+	op, err := a.firestoreAdminClient.UpdateField(ctx, req)
+	if err != nil {
+		return fmt.Errorf("creating FirestoreField %s: %w", fqn, err)
+	}
+
+	created, err := op.Wait(ctx)
+	if err != nil {
+		return fmt.Errorf("FirestoreField %s waiting creation: %w", fqn, err)
+	}
+	log.V(0).Info("successfully created FirestoreField", "name", fqn)
+
+	return a.updateStatus(ctx, createOp, created)
```

Comment:
Nit: There is a race here where we could get the create/update but fail to update the status. Seems like it would be good to mark that the update/create was started before we begin. Makes understanding the status and resuming easier.